shim.c

```diff
@@ -388,6 +392,39 @@ static BOOLEAN verify_x509(UINT8 *Cert, UINTN CertSize)
 	return TRUE;
 }
 
+static BOOLEAN verify_eku(UINT8 *Cert, UINTN CertSize)
+{
+	X509 *x509;
+	CONST UINT8 *Temp = Cert;
+	EXTENDED_KEY_USAGE *eku;
+	ASN1_OBJECT *module_signing;
+
+	module_signing = OBJ_nid2obj(OBJ_create("1.3.6.1.4.1.2312.16.1.2",
+	                                        NULL, NULL));
+
+	x509 = d2i_X509 (NULL, &Temp, (long) CertSize);
+	if (x509 != NULL) {
+		eku = X509_get_ext_d2i(x509, NID_ext_key_usage, NULL, NULL);
+
+		if (eku) {
+			int i = 0;
+			for (i = 0; i < sk_ASN1_OBJECT_num(eku); i++) {
+				ASN1_OBJECT *key_usage = sk_ASN1_OBJECT_value(eku, i);
+
+				if (OBJ_cmp(module_signing, key_usage) == 0)
```

Comment:
There should probably be some discussion around this. It implements a black list where only the 1.3.6.1.4.1.2312.16.1.2 key purpose OID (module signing key) is blocked. The way that I read the [Extended Key Usage section of RFC5280](https://tools.ietf.org/html/rfc5280#section-4.2.1.12) is that if a eku is specified, that the expected eku must be specified in the cert:

>  If the extension is present, then the certificate MUST only be used for one of the purposes indicated.

So if an eku is not specified, then any key will be accepted. However, if an eku is set, I think the RFC reads as if we should implement a whitelist of key purpose OIDs. Reading through the previously submitted [kernel patch set from David Howells](https://lkml.org/lkml/2015/7/17/567) I think that is probably the 1.3.6.1.4.1.2312.16.1.1 key purpose OID (firmware signing key) but I'm not completely sure about that.